win32/tests/gballoc_ll_jemalloc_int/gballoc_ll_jemalloc_int.c

```diff
@@ -260,4 +263,131 @@ TEST_FUNCTION(gballoc_ll_print_stats_works)
+        ASSERT_ARE_EQUAL(int, 0, je_mallctl("thread.arena", NULL, NULL, &arena_id, sizeof(arena_id)));
+
+        ptr[i] = gballoc_ll_malloc(4); // this malloc will happen in the arena i
+        ASSERT_IS_NOT_NULL(ptr[i]);
+    }
+}
+
+TEST_FUNCTION(gballoc_ll_set_option_works_for_dirty_decay)
+{
+    /// arrange
+    // configure the max number of arenas to 4 (hence total max limit for number of arenas = 5, 4 normal + 1 huge)
+    je_malloc_conf = "narenas:4";
+
+    uint32_t expected_narenas = MAX_ARENAS;
+    void* ptr[MAX_ARENAS] = { NULL };
+
+    setup_arenas_for_set_option(expected_narenas, ptr);
+
+    int result;
+    int64_t decay_ms = DECAY_MS;
+    size_t decay_ms_size = sizeof(decay_ms);
+    int64_t verify_decay_ms;
+    int64_t old_decay_ms;
+
+    // Retrieve the old decay value
+    ASSERT_ARE_EQUAL(int, 0, je_mallctl("arenas.dirty_decay_ms", &old_decay_ms, &decay_ms_size, NULL, 0));
+    ASSERT_ARE_NOT_EQUAL(int64_t, decay_ms, old_decay_ms);
+
+    ///act
+    result = gballoc_ll_set_option("dirty_decay", &decay_ms);
+
+    ///assert - doesn't crash
+    ASSERT_ARE_EQUAL(int, 0, result);
+    ASSERT_ARE_EQUAL(int, 0, je_mallctl("arenas.dirty_decay_ms", &verify_decay_ms, &decay_ms_size, NULL, 0));
+    ASSERT_ARE_EQUAL(int64_t, decay_ms, verify_decay_ms);
+
+    // verify that all the arenas have the decay set except arena 1 and last arena(reserved for huge arena)
+    assert_set_option_works_for_decay("dirty_decay", decay_ms, expected_narenas);
+
+    ///clean
+    for (uint32_t i = 0; i < expected_narenas; i++)
+    {
+        gballoc_ll_free(ptr[i]);
+    }
+}
+
+TEST_FUNCTION(gballoc_ll_set_option_works_for_muzzy_decay)
+{
+    /// arrange
+    // configure the max number of arenas to 4 (hence total max limit for number of arenas = 5, 4 normal + 1 huge)
+    je_malloc_conf = "narenas:4";
+
+    uint32_t expected_narenas = MAX_ARENAS;
+    void* ptr[MAX_ARENAS] = { NULL };
+
+    setup_arenas_for_set_option(expected_narenas, ptr);
+
+    int result;
+    int64_t decay_ms = DECAY_MS;
+    size_t decay_ms_size = sizeof(decay_ms);
+    int64_t verify_decay_ms;
+    int64_t old_decay_ms;
+
+    // Retrieve the old decay value
+    ASSERT_ARE_EQUAL(int, 0, je_mallctl("arenas.muzzy_decay_ms", &old_decay_ms, &decay_ms_size, NULL, 0));
+    ASSERT_ARE_NOT_EQUAL(int64_t, decay_ms, old_decay_ms);
+
+    ///act
+    result = gballoc_ll_set_option("muzzy_decay", &decay_ms);
+
+    ///assert - doesn't crash
+    ASSERT_ARE_EQUAL(int, 0, result);
+    ASSERT_ARE_EQUAL(int, 0, je_mallctl("arenas.muzzy_decay_ms", &verify_decay_ms, &decay_ms_size, NULL, 0));
+    ASSERT_ARE_EQUAL(int64_t, decay_ms, verify_decay_ms);
+
+    // verify that all the arenas have the decay set except arena 1 and last arena(reserved for huge arena)
+    assert_set_option_works_for_decay("muzzy_decay", decay_ms, expected_narenas);
+
+    ///clean
+    for (uint32_t i = 0; i < expected_narenas; i++)
+    {
+        gballoc_ll_free(ptr[i]);
+    }
+}
```

Comment:
I think we're missing tests here, and it is better to have them now and here (in c-pal) rather than to discover that they don't work later and in production.

We should have a test that actually sees when memory is released to OS and depending on the config, it is released earlier/later/never at all.

I can imagine something like:
let's start like with a number of threads, all of them pretending to do memory work (alloc/free) for an amount of time.

Following this amount of time, some memory should still be used, some memory should be free(). Take a snapshot of the memory used by the process.

Wait past the decay timer (there's no other activity)

Take another snapshot of the memory used by the process.

The diff between the 2 snapshots and the decay parameter should be correlated as a condition to pass the test. With some tolerances, ofc.

Reply:
I still think this PR requires at least the above work to be completed, so I'll still be waiting. Rest looks good.

Reply:
I have added new tests that check the dirty and muzzy page numbers after we decay the arenas. Hope these would be sufficient to test the changes?

Reply:
I looked at the tests, and while they do seem to test what I've mentioned here, it is a case of "testing jemalloc with jemalloc".

When I said before "Take a snapshot of the memory used by the process." I was referring to the working set size. So let's compare the working set "before" and "after" by calling Windows APIs and not use jemalloc's internals.

Reply:
Need to see the next version of the INT tests with Windows.

I understand if there are no INT tests for Linux at this time, but there should be some INT tests for Linux too using the parallel mechanisms existing for Linux.

Reply:
Done